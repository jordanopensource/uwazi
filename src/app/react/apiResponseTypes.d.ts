import { UserGroupSchema } from 'shared/types/userGroupType';
import { UserSchema } from 'shared/types/userType';
import {
  PreserveConfig,
  Settings,
  SettingsFilterSchema,
  SettingsLinkSchema,
} from 'shared/types/settingsType';
import { LanguageSchema } from 'shared/types/commonTypes';
import { ThesaurusSchema, ThesaurusValueSchema } from 'shared/types/thesaurusType';
import { TemplateSchema } from 'shared/types/templateType';

export interface GroupMemberSchema {
  refId: string;
  username: string;
}

export interface ClientUserGroupSchema extends Omit<UserGroupSchema, '_id' | 'members'> {
  _id?: string;
  members: GroupMemberSchema[];
}

export interface ClientUserSchema extends Omit<UserSchema, '_id' | 'groups'> {
  _id?: string;
  groups?: {
    _id: string;
    name: string;
  }[];
}

export interface ClientSettingsFilterSchema extends Omit<SettingsFilterSchema, '_id'> {
  _id?: string;
}

export interface ClientSublink extends Omit<ClientSettingsLinkSchema, 'sublinks'> {}

export interface ClientSettingsLinkSchema extends Omit<SettingsLinkSchema, '_id' | 'sublinks'> {
  _id?: string;
  sublinks?: ClientSublink[];
}

export interface ClientPreserveConfig extends Omit<PreserveConfig, 'config'> {
  config: {
    token: string;
    template: string;
    user?: string;
  }[];
}

export interface ClientLanguageSchema extends Omit<LanguageSchema, '_id'> {
  _id?: string;
}

export interface Template extends TemplateSchema, Omit<TemplateSchema, '_id'> {
  _id: string;
}

export interface ClientSettings
  extends Omit<Settings, '_id | filters | links | features | languages'> {
  _id?: string;
  filters?: ClientSettingsFilterSchema[];
  languages?: ClientLanguageSchema[];
  links?: ClientSettingsLinkSchema[];
  features?: Omit<Settings['features'], 'preserve | metadataExtraction'> & {
    preserve?: ClientPreserveConfig;
    ocr?: { url: string };
    metadataExtraction?: {
      url: string;
      templates?: {
        template: string;
        properties: string[];
      }[];
    };
    paragraphExtraction?: boolean;
    [k: string]: unknown | undefined;
  };
}

export interface ClientThesaurus extends ThesaurusSchema, Omit<ThesaurusSchema, '_id | values'> {
  _id: ObjectIdSchema;
  values: ClientThesaurusValue[];
}

export interface ClientThesaurusValue extends Omit<ThesaurusValueSchema, '_id'> {
  id?: string;
  label: string;
  values?: {
    id?: string;
    label: string;
    name?: string;
  }[];
}

export interface ClientRelationshipType {
  _id: string;
  name: string;
}
