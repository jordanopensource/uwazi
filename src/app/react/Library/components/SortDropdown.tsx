import React, { useRef, useState, useCallback, useEffect } from 'react';
import { bindActionCreators, Dispatch } from 'redux';
import { connect, ConnectedProps } from 'react-redux';
import { actions } from 'react-redux-form';
import { risonDecodeOrIgnore } from 'app/utils';
import { useLocation, useSearchParams } from 'react-router-dom';
import { Icon } from 'UI';
import { I18NLink, t } from 'app/I18N';
import { wrapDispatch } from 'app/Multireducer';
import { IStore } from 'app/istore';
import { IImmutable } from 'shared/types/Immutable';
import { useOnClickOutsideElement } from 'app/utils/useOnClickOutsideElementHook';
import { encodeSearch } from '../actions/libraryActions';
import {
  getCurrentSortOption,
  getPropertySortType,
  SearchOptions,
  SortType,
  filterTemplates,
  getSortOptions,
} from '../helpers/sortComponets';

interface SortDropdownOwnProps {
  selectedTemplates: IImmutable<string[]>;
}

const getOptionUrl = (option: SortType, path: string, query: string | null) => {
  const currentQuery = risonDecodeOrIgnore(decodeURIComponent(query || '()'));
  const type = getPropertySortType(option);
  return `${path}${encodeSearch(
    { ...currentQuery, order: type === 'string' ? 'asc' : 'desc', sort: option.value, from: 0 },
    true
  )}`;
};

const mapStateToProps = (state: IStore, ownProps: SortDropdownOwnProps) => {
  let { templates } = state;

  if (ownProps.selectedTemplates && ownProps.selectedTemplates.count()) {
    templates = filterTemplates(state.templates, ownProps.selectedTemplates);
  }

  return {
    templates,
    locale: state.locale,
  };
};

const mapDispatchToProps = (dispatch: Dispatch<{}>) =>
  bindActionCreators({ merge: actions.merge }, wrapDispatch(dispatch, 'library'));

const connector = connect(mapStateToProps, mapDispatchToProps);

type mappedProps = ConnectedProps<typeof connector>;

// eslint-disable-next-line max-statements
const SortDropdownComponent = ({ templates, locale }: mappedProps) => {
  const [dropdownOpen, setDropdownOpen] = useState(false);
  const menuRef = useRef(null);
  const location = useLocation();
  const [searchParams] = useSearchParams();
  const currentQuery: SearchOptions = risonDecodeOrIgnore(
    decodeURIComponent(searchParams.get('q') || '()')
  );
  const path = location.pathname.replace(new RegExp(`^/?${locale}/|^/?${locale}$`), '');
  const sortButtonLink = `${path}${encodeSearch(
    { ...currentQuery, order: currentQuery.order === 'asc' ? 'desc' : 'asc', from: 0 },
    true
  )}`;

  useEffect(() => {
    setDropdownOpen(false);
  }, [currentQuery.sort]);

  useOnClickOutsideElement<HTMLDivElement>(
    menuRef,
    useCallback(() => {
      setDropdownOpen(false);
    }, [])
  );

  const sortOptions = getSortOptions(currentQuery, templates);

  return (
    <div className="sort-buttons">
      <div className="sort-dropdown" ref={menuRef}>
        <button
          type="button"
          className={`dropdown-button ${dropdownOpen ? 'expanded' : ''}`}
          onClick={() => setDropdownOpen(!dropdownOpen)}
        >
          <span>{getCurrentSortOption(sortOptions, currentQuery.sort)}</span>
          <Icon icon={dropdownOpen ? 'caret-up' : 'caret-down'} />
        </button>

        <ul className={`dropdown-menu ${dropdownOpen ? 'expanded' : ''}`}>
          {sortOptions.map(option => {
            const url = getOptionUrl(option, path, searchParams.get('q'));
            return (
              <li key={option.value}>
                <I18NLink to={url} href={url}>
                  {option.label}
                </I18NLink>
              </li>
            );
          })}
        </ul>
      </div>

      <I18NLink
        to={sortButtonLink}
        href={sortButtonLink}
        disable={currentQuery.sort === '_score' ? 'true' : undefined}
      >
        <button
          type="button"
          disabled={currentQuery.sort === '_score' ? true : undefined}
          onClick={() => {}}
          className="sorting-toggle"
        >
          <span style={{ display: 'none' }}>
            {currentQuery.order === 'asc'
              ? t('System', 'Sort descending', undefined, false)
              : t('System', 'Sort ascending', undefined, false)}
          </span>
          <Icon
            icon={
              currentQuery.order === 'asc' && currentQuery.sort !== '_score'
                ? 'arrow-up'
                : 'arrow-down'
            }
          />
        </button>
      </I18NLink>
    </div>
  );
};

const container = connector(SortDropdownComponent);
export { container as SortDropdown };
