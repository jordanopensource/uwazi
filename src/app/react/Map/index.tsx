/* eslint-disable react/jsx-props-no-spreading */
import loadable from '@loadable/component';
import React from 'react';
import { Map, Layer } from 'app/Map/MapContainer';
import * as helper from './helper';

const LMap = loadable(async () => {
  const { LMap: LMapComponent } = await import(/* webpackChunkName: "LazyLoadMap" */ './LMap');
  return (props: any) => <LMapComponent {...props} />;
});

export { default as Markers } from './Markers';
export { helper, LMap, Map };
export type { Layer };
