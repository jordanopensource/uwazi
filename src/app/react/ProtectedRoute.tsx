import React, { ReactElement } from 'react';
import { Navigate, Outlet } from 'react-router-dom';
import { store } from 'app/store';
import { ClientSettings } from 'app/apiResponseTypes';

const ProtectedRoute = ({
  children,
  allowedRoles,
}: {
  children: ReactElement;
  allowedRoles?: string[];
}) => {
  const userId = store?.getState().user.get('_id');
  const userRole = store?.getState().user.get('role') || '';
  if (allowedRoles && allowedRoles.includes(userRole)) {
    return children || <Outlet />;
  }

  if (!allowedRoles && userId) {
    return children || <Outlet />;
  }

  return <Navigate to="/login" replace />;
};

const adminsOnlyRoute = (element: ReactElement) => (
  <ProtectedRoute allowedRoles={['admin']}>{element}</ProtectedRoute>
);

const privateRoute = (element: ReactElement, settings: ClientSettings | undefined) =>
  !settings?.private ? element : <ProtectedRoute>{element}</ProtectedRoute>;

const loggedInUsersRoute = (element: ReactElement) => <ProtectedRoute>{element}</ProtectedRoute>;
export { loggedInUsersRoute, adminsOnlyRoute, privateRoute, ProtectedRoute };
