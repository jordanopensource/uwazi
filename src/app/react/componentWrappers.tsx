/* eslint-disable comma-spacing */
/* eslint-disable react/no-multi-comp */
import React, { useContext } from 'react';
import {
  useLocation,
  useMatches,
  useNavigate,
  useOutlet,
  useParams,
  useSearchParams,
} from 'react-router-dom';
import { AppMainContext } from './App/AppMainContext';
import { searchParamsFromSearchParams } from './utils/routeHelpers';

const withRouter =
  <T,>(Component: React.ComponentClass<T & { params?: any }, any>) =>
  (props: T & { params?: any }) => {
    const location = useLocation();
    const navigate = useNavigate();
    const params = useParams();
    const matches = useMatches();
    const [searchParams, setSearchParams] = useSearchParams();
    const { q: searchString } = searchParamsFromSearchParams(searchParams);
    const paramsFromProps = { ...props.params, q: searchString || props.params?.q };

    return (
      <Component
        {...props}
        navigate={navigate}
        location={location}
        params={{ ...params, ...(paramsFromProps || {}) }}
        matches={matches}
        searchParams={searchParams}
        setSearchParams={setSearchParams}
      />
    );
  };

const withContext =
  <T,>(Component: React.ComponentClass<T, any>) =>
  (props: T) => {
    const mainContext = useContext(AppMainContext);
    return <Component {...props} mainContext={mainContext} />;
  };

const withOutlet =
  <T,>(Component: React.ComponentClass<T, any>) =>
  (props: T) => {
    const outlet = useOutlet(AppMainContext);
    return <Component {...props} outlet={outlet} />;
  };

export { withRouter, withContext, withOutlet };
