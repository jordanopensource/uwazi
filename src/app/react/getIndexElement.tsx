import React from 'react';
import { Navigate } from 'react-router-dom';
import { ClientSettings } from 'app/apiResponseTypes';
import { validateHomePageRoute } from './utils/routeHelpers';
import { PageView } from './Pages/PageView';
import { LibraryTable } from './Library/LibraryTable';
import { LibraryMap } from './Library/LibraryMap';
import { LibraryCards } from './Library/Library';
import { Login } from './Users/Login';
import { ViewerRoute } from './Viewer/ViewerRoute';

const deconstructSearchQuery = (query?: string) => {
  if (!query) return '';
  if (query.startsWith('?q=')) {
    return decodeURI(query.substring(1).split('=')[1]);
  }
  return `(${query.substring(1)})`;
};

const getCustomLibraryPage = (customHomePage: string[]) => {
  const [query] = customHomePage.filter(path => path.startsWith('?'));
  const searchQuery = deconstructSearchQuery(query);
  const queryString = query ? searchQuery : '';

  if (customHomePage.includes('map')) {
    return <LibraryMap params={{ q: queryString }} />;
  }

  if (customHomePage.includes('table')) {
    return <LibraryTable params={{ q: queryString }} />;
  }

  return <LibraryCards params={{ q: queryString }} />;
};

const getLibraryDefault = (
  userId: string | undefined,
  defaultLibraryView: string | undefined,
  privateInstance: boolean | undefined
) => {
  if (userId) {
    return <Navigate to="/library/?q=(includeUnpublished:!t)" state={{ isClient: true }} />;
  }

  if (privateInstance) {
    return <Login />;
  }

  switch (defaultLibraryView) {
    case 'table':
      return <LibraryTable />;

    case 'map':
      return <LibraryMap />;

    case 'cards':
    default:
      return <LibraryCards />;
  }
};

// eslint-disable-next-line max-statements
const getIndexElement = (settings: ClientSettings | undefined, userId: string | undefined) => {
  const customHomePage = settings?.home_page ? settings?.home_page.split('/').filter(v => v) : [];
  const isValidHomePage = validateHomePageRoute(settings?.home_page || '');
  let element = <Navigate to={customHomePage.join('/')} />;
  let parameters;

  switch (true) {
    case !isValidHomePage || customHomePage.length === 0:
      element = getLibraryDefault(userId, settings?.defaultLibraryView, settings?.private);
      break;

    case isValidHomePage && customHomePage.includes('page'):
      {
        const pageId = customHomePage[customHomePage.indexOf('page') + 1];
        element = <PageView params={{ sharedId: pageId }} />;
        parameters = { sharedId: pageId };
      }
      break;

    case isValidHomePage && customHomePage.includes('entity'):
      {
        const pageId = customHomePage[customHomePage.indexOf('entity') + 1];
        element = <ViewerRoute params={{ sharedId: pageId }} />;
      }
      break;

    case isValidHomePage && customHomePage.includes('library'):
      element = getCustomLibraryPage(customHomePage);
      break;

    default:
      break;
  }

  return { element, parameters };
};

export { getIndexElement };
