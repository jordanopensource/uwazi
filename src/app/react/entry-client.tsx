import React from 'react';
import { hydrateRoot } from 'react-dom/client';
import * as Sentry from '@sentry/react';
import {
  RouterProvider,
  createBrowserRouter,
  createRoutesFromChildren,
  matchRoutes,
  useLocation,
  useNavigationType,
} from 'react-router-dom';
import { Provider } from 'jotai';
import { Provider as ReduxProvider } from 'react-redux';
import type { RequestError } from 'V2/shared/errorUtils';
import { ErrorBoundary } from './V2/Components/ErrorHandling';
import './App/sockets';
import { getRoutes } from './Routes';
import CustomProvider from './App/Provider';
import { settingsAtom, atomStore, userAtom } from './V2/atoms';
import { store } from './store';

declare global {
  interface Window {
    __loadingError__?: RequestError;
  }
}

if (window.SENTRY_APP_DSN) {
  Sentry.init({
    release: window.UWAZI_VERSION,
    environment: window.UWAZI_ENVIRONMENT,
    dsn: window.SENTRY_APP_DSN,
    integrations: [
      new Sentry.BrowserTracing({
        routingInstrumentation: Sentry.reactRouterV6Instrumentation(
          React.useEffect,
          useLocation,
          useNavigationType,
          createRoutesFromChildren,
          matchRoutes
        ),
      }),
      new Sentry.Replay(),
    ],

    tracesSampleRate: 0.1,
  });
}

const router = createBrowserRouter(
  getRoutes(atomStore.get(settingsAtom), atomStore.get(userAtom)?._id)
);

const App = () => (
  <ReduxProvider store={store as any}>
    <CustomProvider>
      <Provider store={atomStore}>
        <ErrorBoundary>
          <RouterProvider router={router} fallbackElement={null} />
        </ErrorBoundary>
      </Provider>
    </CustomProvider>
  </ReduxProvider>
);

const container = document.getElementById('root');
const root = window.__loadingError__ === undefined ? hydrateRoot(container!, <App />) : container;
const silentWarnings = [
  'Warning: %s uses the legacy childContextTypes API which is no longer supported and will be removed in the next major release.',
  'Warning: %s: Support for defaultProps will be removed from function components in a future major release. Use JavaScript default parameters instead.%s',
  'Warning: %s uses the legacy contextTypes API which is no longer supported and will be removed in the next major release.',
  'Warning: findDOMNode is deprecated and will be removed in the next major release.',
];

const isSilentWarning = (warning: any) =>
  silentWarnings.find(w => typeof warning === 'string' && warning.includes(w)) !== undefined;

const origConsoleError = window.console.error;

window.console.error = (...args) => {
  if (isSilentWarning(args[0])) {
    return;
  }
  origConsoleError.apply(window.console, args);
};

export { root };
