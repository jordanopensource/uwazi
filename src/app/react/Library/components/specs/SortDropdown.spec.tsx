/**
 * @jest-environment jsdom
 */
import React from 'react';
import { fromJS } from 'immutable';
import { RenderResult, screen } from '@testing-library/react';
import { defaultState, renderConnectedContainer } from 'app/utils/test/renderConnected';
import { SortDropdown } from '../SortDropdown';

describe('Sort dropdown', () => {
  let props: any;
  let renderResult: RenderResult;
  let search: string;

  const templates = [
    {
      _id: 'id',
      properties: [
        {},
        { filter: true, name: 'date', label: 'Date', type: 'date' },
        { filter: true, name: 'number', label: 'Number', type: 'numeric' },
        { filter: true, name: 'my_select', label: 'My select', type: 'select' },
        { filter: true, name: 'sortable_name', label: 'Sortable name', type: 'text' },
        {
          filter: true,
          name: 'inherited_1',
          label: 'Inherited 1',
          type: 'relationship',
          inherit: {
            type: 'text',
          },
        },
        {
          filter: true,
          name: 'non_sortable_property',
          label: 'Non sortable property',
          type: 'relationship',
          inherit: {
            type: 'multiselect',
          },
        },
      ],
    },
  ];

  beforeEach(() => {
    props = {
      locale: 'en',
      templates: fromJS(templates),
      search: {
        order: 'asc',
        sort: 'creationDate',
        searchTerm: '',
      },
    };
    search = '(from:0,includeUnpublished:!t,limit:30,order:desc)';
  });

  const render = () => {
    ({ renderResult } = renderConnectedContainer(
      <SortDropdown.WrappedComponent {...props} />,
      () => ({
        ...defaultState,
      }),
      'MemoryRouter',
      [`/en/library/?q=${search}`]
    ));
  };

  it('should display sorting by creation date by default and have that option in the sort list', () => {
    render();
    const [sortButton, listOption] = screen.getAllByText('Date added');
    expect(sortButton.parentElement).toHaveClass('dropdown-button');
    expect(listOption.closest('a')).toHaveAttribute(
      'href',
      '/en/library/?q=(from:0,includeUnpublished:!t,limit:30,order:desc,sort:creationDate)'
    );
  });

  it.each`
    option             | link
    ${'Title'}         | ${'/en/library/?q=(from:0,includeUnpublished:!t,limit:30,order:asc,sort:title)'}
    ${'Date modified'} | ${'/en/library/?q=(from:0,includeUnpublished:!t,limit:30,order:desc,sort:editDate)'}
    ${'Date'}          | ${'/en/library/?q=(from:0,includeUnpublished:!t,limit:30,order:desc,sort:metadata.date)'}
    ${'Number'}        | ${'/en/library/?q=(from:0,includeUnpublished:!t,limit:30,order:desc,sort:metadata.number)'}
    ${'My select'}     | ${'/en/library/?q=(from:0,includeUnpublished:!t,limit:30,order:asc,sort:metadata.my_select)'}
    ${'Sortable name'} | ${'/en/library/?q=(from:0,includeUnpublished:!t,limit:30,order:asc,sort:metadata.sortable_name)'}
    ${'Inherited 1'}   | ${'/en/library/?q=(from:0,includeUnpublished:!t,limit:30,order:desc,sort:metadata.inherited_1.inheritedValue)'}
  `(
    'should display the sortable option $option with the link, and correct sort and order',
    ({ option, link }) => {
      render();
      expect(screen.getByText(option).closest('a')).toHaveAttribute('href', link);
    }
  );

  it('should not display properties that are not sortable', () => {
    render();
    expect(screen.queryByText('Non sortable property')).not.toBeInTheDocument();
  });

  it('should not duplicate the entry when multiple options have the same property name', () => {
    props.templates = fromJS([
      {
        _id: 'id1',
        properties: [
          {
            filter: true,
            name: 'sortable_text',
            label: 'My text',
            type: 'text',
            _id: '_id',
          },
        ],
      },
      {
        _id: 'id2',
        properties: [
          {
            filter: true,
            name: 'sortable_text',
            label: 'My text',
            type: 'text',
            _id: '_id',
          },
        ],
      },
    ]);

    render();
    const listOption = screen.getAllByText('My text');
    expect(listOption.length).toBe(1);
  });

  describe('when there is an active search term', () => {
    it('should display sort by search relevance option in the button and in the list, and disable the sort button', () => {
      search = "(searchTerm:'my search',sort:_score)";
      render();
      const byRelevanceText = screen.getAllByText('Search relevance');
      expect(byRelevanceText).toHaveLength(2);
      expect(renderResult.container).toMatchSnapshot();
    });
  });

  describe('sort button', () => {
    it('should reverte the sorting order and keep other parameters as they are', () => {
      render();
      expect(screen.getByText('Sort ascending').parentElement?.parentElement).toHaveAttribute(
        'href',
        '/en/library/?q=(from:0,includeUnpublished:!t,limit:30,order:asc)'
      );

      search = "(from:0,includeUnpublished:!t,limit:30,order:asc,searchTerm:'some search')";
      render();
      expect(screen.getByText('Sort descending').parentElement?.parentElement).toHaveAttribute(
        'href',
        "/en/library/?q=(from:0,includeUnpublished:!t,limit:30,order:desc,searchTerm:'some search')"
      );
    });
  });
});
