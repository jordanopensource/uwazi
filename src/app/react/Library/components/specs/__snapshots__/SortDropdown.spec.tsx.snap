// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort dropdown when there is an active search term should display sort by search relevance option in the button and in the list, and disable the sort button 1`] = `
<div>
  <div
    class="sort-buttons"
  >
    <div
      class="sort-dropdown"
    >
      <button
        class="dropdown-button "
        type="button"
      >
        <span>
          Search relevance
        </span>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-down fa-w-10 "
          data-icon="caret-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            fill="currentColor"
          />
        </svg>
      </button>
      <ul
        class="dropdown-menu "
      >
        <li>
          <a
            aria-current="page"
            href="/en/library/?q=(from:0,order:asc,searchTerm:'my search',sort:title)"
          >
            Title
          </a>
        </li>
        <li>
          <a
            aria-current="page"
            href="/en/library/?q=(from:0,order:desc,searchTerm:'my search',sort:creationDate)"
          >
            Date added
          </a>
        </li>
        <li>
          <a
            aria-current="page"
            href="/en/library/?q=(from:0,order:desc,searchTerm:'my search',sort:editDate)"
          >
            Date modified
          </a>
        </li>
        <li>
          <a
            aria-current="page"
            href="/en/library/?q=(from:0,order:desc,searchTerm:'my search',sort:_score)"
          >
            Search relevance
          </a>
        </li>
        <li>
          <a
            aria-current="page"
            href="/en/library/?q=(from:0,order:desc,searchTerm:'my search',sort:metadata.date)"
          >
            Date
          </a>
        </li>
        <li>
          <a
            aria-current="page"
            href="/en/library/?q=(from:0,order:desc,searchTerm:'my search',sort:metadata.number)"
          >
            Number
          </a>
        </li>
        <li>
          <a
            aria-current="page"
            href="/en/library/?q=(from:0,order:asc,searchTerm:'my search',sort:metadata.my_select)"
          >
            My select
          </a>
        </li>
        <li>
          <a
            aria-current="page"
            href="/en/library/?q=(from:0,order:asc,searchTerm:'my search',sort:metadata.sortable_name)"
          >
            Sortable name
          </a>
        </li>
        <li>
          <a
            aria-current="page"
            href="/en/library/?q=(from:0,order:desc,searchTerm:'my search',sort:metadata.inherited_1.inheritedValue)"
          >
            Inherited 1
          </a>
        </li>
      </ul>
    </div>
    <a
      aria-current="page"
      disable="true"
      href="/en/library/?q=(from:0,order:asc,searchTerm:'my search',sort:_score)"
    >
      <button
        class="sorting-toggle"
        disabled=""
        type="button"
      >
        <span
          style="display: none;"
        >
          Sort ascending
        </span>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-arrow-down fa-w-14 "
          data-icon="arrow-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
            fill="currentColor"
          />
        </svg>
      </button>
    </a>
  </div>
</div>
`;
